mp/src/game/shared/momentum/util/engine_patch.cpp

```diff
@@ -0,0 +1,243 @@
+//------------------------------------------------------------------------------------
+// Functions used to find patterns of bytes in the engine's memory to hook or patch
+//------------------------------------------------------------------------------------
+#ifdef _WIN32
+#define WIN32_LEAN_AND_MEAN
+#include "Windows.h"
+#include "Psapi.h"
+#pragma comment(lib, "psapi.lib")
+#elif defined (POSIX)
+#include "util/os_utils.h"
+#include "sys/mman.h"
+#endif
+
+#include "engine_patch.h"
+#include "mom_shareddefs.h"
+#include "tier0/platform.h"
+
+void* CEngineBinary::moduleBase = nullptr;
+size_t CEngineBinary::moduleSize;
+
+CEngineBinary::CEngineBinary() : CAutoGameSystem("CEngineBinary")
+{
+}
+
+// Get the engine's base address and size
+bool CEngineBinary::Init()
+{
+#ifdef _WIN32
+    HMODULE handle = GetModuleHandleA("engine.dll");
+    if (!handle)
+        return false;
+
+    MODULEINFO info;
+    GetModuleInformation(GetCurrentProcess(), handle, &info, sizeof(info));
+
+    moduleBase = info.lpBaseOfDll;
+    moduleSize = info.SizeOfImage;
+#else //POSIX
+    if (GetModuleInformation(ENGINE_DLL_NAME, &moduleBase, &moduleSize))
+        return false;
+#endif //WIN32
+
+    return true;
+}
+
+void CEngineBinary::PostInit()
+{
+    CEnginePatch::ApplyAll();
+}
+
+inline bool CEngineBinary::DataCompare(const unsigned char* data, const unsigned char* pattern, const char* mask)
+{
+    for (; *mask != 0; ++data, ++pattern, ++mask)
+        if (*mask == 'x' && *data != *pattern)
+            return false;
+
+    return (*mask == 0);
+}
+
+//---------------------------------------------------------------------------------------------------------
+// Finds a pattern of bytes in the engine memory given a signature and a mask
+// Returns the address of the first (and hopefully only) match with an optional offset, otherwise nullptr
+//---------------------------------------------------------------------------------------------------------
+void* CEngineBinary::FindPattern(const unsigned char* pattern, const char* mask, size_t offset)
+{
+    auto maskLength = strlen(mask);
+    for (size_t i = 0; i <= moduleSize - maskLength; ++i)
+    {
+        auto addr = reinterpret_cast<const unsigned char*>(moduleBase) + i;
```

Comment:
`moduleBase` isn't a a `const` variable, so seems kind of redundant to me that you're adding the `const` here only to `const_cast` it out a few lines later.

Reply:
Most of the code was moved from tickset so I didn't notice, good catch.